### DISCUSSION TITLE
New Ubuntu install - nothing loads through VPN

### DISCUSSION THREAD

**Discussion by joseb552** [pivpn/pivpn#1845] [Support]
I've had a PiVPN instance running on a raspberry pi for a few years and that worked perfectly. Recently I've been having some issues that I think might be related to the Pi itself, so to test it I decided to set up a new PiHole / PiVPN system on an HP mini-PC. I installed Ubuntu 24.04 LTS Desktop and was able to get PiHole installed and configured. 

I then installed PiVPN, created a profile and configured my router to forward to the new HP / Ubuntu install. From my iPhone I can connect to the VPN server and if I run 'pivpn -c' on the terminal I can see that the profile is active and it does show some bytes sent and received but nothing will load on the phone. I've looked through some tutorials for any mention of firewall changes needed but haven't found anything. 

Is there additional setup needed for this? I do want all VPN traffic to flow through the PiHole, maybe there's something there that needs to be changed? When I set up the PiVPN it did recognize that PiHole was installed and I told it I did want to use PiHole / PiVPN together. 

 I ran pivpn -d and it did find a couple of 'issues' that I told it to correct. After that I rebooted the PC and tried again but still nothing for internet loads. I CAN connect to local devices (PiHole admin interface) while on VPN, just nothing on the internet. I did have to change one setting on the PiHole config from what I normally have. In Settings > DNS, I had to change from "Allow only local requests" to "Respond Only On Interface enp2s0" because I was not able to browse the internet with the initial setting. On my Pi PiHole I never had to change that setting. 

```
    ::::		Self check		 ::::
    :: [OK] IP forwarding is enabled
    :: [ERR] Iptables MASQUERADE rule is not set, attempt fix now? [Y/n] y
    Done
    :: [ERR] Iptables INPUT rule is not set, attempt fix now? [Y/n] y
    Done
    :: [OK] WireGuard is running
    :: [OK] WireGuard is enabled
    (it will automatically start on reboot)
    :: [OK] WireGuard is listening on port 56580/udp
    [INFO] Run pivpn -d again to see if we detect issues
```

**Comment by orazioedoardo** [comment]
Post the full output of `pivpn debug`

**Reply by joseb552** [reply]
Created a new gist for it here: 

https://gist.github.com/joseb552/5c49bafe489c4ef55ae17043b08ee236

**Reply by mwoolweaver** [reply]
> [ERR] Iptables MASQUERADE rule is not set, attempt fix now? [Y/n] Done
Done

is pivpn able to fix this error?? or does that error show up everytime you run `pivpn -d`??

**Reply by joseb552** [reply]
It is able to correct it, if I re-run diagnostics it shows that it is OK. However, if the system is rebooted and I try diagnostics again, it shows that it needs to be corrected.

**Reply by mwoolweaver** [reply]
can you run `sudo dpkg -S iptables-persistent` or `dpkg -l iptables-persistent` and post the output if any is given?

**Reply by joseb552** [reply]
```
odge@odge-pihole: sudo dpkg -S iptables-persistent
dpkg-query: no path found matching pattern *iptables-persistent*
odge@odge-pihole: dpkg -l iptables-persistent
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                Version      Architecture Description
+++-===================-============-============-=====================================================
rc  iptables-persistent 1.0.20       all          boot-time loader for netfilter rules, iptables plugin
```

**Reply by mwoolweaver** [reply]
it seems that iptables-persistent is not installed completely (or something) based on the output of those 2 commands when i run them

```
 vpn@vpn:\~$ sudo dpkg -S iptables-persistent
 iptables-persistent: /usr/share/doc/iptables-persistent/changelog.gz
 iptables-persistent: /usr/share/doc/iptables-persistent/README
 iptables-persistent: /usr/share/doc/iptables-persistent/NEWS.Debian.gz
 iptables-persistent: /usr/share/doc/iptables-persistent
 iptables-persistent: /usr/share/doc/iptables-persistent/copyright
 vpn@vpn:\~$ dpkg -l iptables-persistent
 Desired=Unknown/Install/Remove/Purge/Hold
 | Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
 |/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
 ||/ Name                Version      Architecture Description
 +++-===================-============-============-=====================================================
 ii  iptables-persistent 1.0.20       all          boot-time loader for netfilter rules, iptables plugin
```

`apt update` & `apt upgrade` might help

**Reply by mwoolweaver** [reply]
> It is able to correct it, if I re-run diagnostics it shows that it is OK. However, if the system is rebooted and I try diagnostics again, it shows that it needs to be corrected.

does traffic start working after the debug script fixes the iptables?

**Reply by joseb552** [reply]
Thanks for being patient with me, I have a lot more time today to work on this. I tried `apt update` & `apt upgrade `then rebooted the PC. When it came back up I ran `pivpn -d` again and it still was asking to correct the two same issues as before.

**Reply by orazioedoardo** [reply]
As @mwoolweaver said, looks like you are missing `iptables-persistent` which is weird, I just tried installing PiVPN on Ubuntu 24.04 and it works. Run `sudo apt install iptables-persistent` and reboot.

**Reply by joseb552** [reply]
> As @mwoolweaver said, looks like you are missing `iptables-persistent` which is weird, I just tried installing PiVPN on Ubuntu 24.04 and it works. Run `sudo apt install iptables-persistent` and reboot.

I ran `sudo apt install iptables-persistent` then I rebooted the system. When it came back up I ran `pivpn -d` and it returned no issues at all. Tested VPN just now and appears to be working fine (though it has been at least since the last reboot of the host). I wonder if I messed up the install and that caused iptables-persistent to not be installed.

**Comment by orazioedoardo** [comment]
What’s the output of `sudo iptables -vnL -t nat` ?

**Reply by joseb552** [reply]
```
odge@odge-pihole:~$ sudo iptables -vnL -t nat
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain POSTROUTING (policy ACCEPT 7 packets, 566 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 MASQUERADE  0    --  *      enp2s0  10.66.27.0/24        0.0.0.0/0            /* wireguard-nat-rule */

```

**Reply by orazioedoardo** [reply]
Well, it’s correct. What the output of `sudo iptables -vnL` and `sudo cat /etc/iptables/rules.v4` ?

**Reply by joseb552** [reply]
```
odge@odge-pihole:/etc/dnsmasq.d$ sudo iptables -vnL
[sudo] password for odge: 
Chain INPUT (policy ACCEPT 30051 packets, 3707K bytes)
 pkts bytes target     prot opt in     out     source               destination         
 9448 1507K ACCEPT     17   --  enp2s0 *       0.0.0.0/0            0.0.0.0/0            udp dpt:59520 /* wireguard-input-rule */

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
```

and 

```
odge@odge-pihole:/etc/dnsmasq.d$ sudo cat /etc/iptables/rules.v4
# Generated by iptables-save v1.8.10 (nf_tables) on Wed Jun 19 08:14:13 2024
*filter
:INPUT ACCEPT [0:0]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
-A INPUT -i enp2s0 -p udp -m udp --dport 59520 -m comment --comment wireguard-input-rule -j ACCEPT
COMMIT
# Completed on Wed Jun 19 08:14:13 2024
# Generated by iptables-save v1.8.10 (nf_tables) on Wed Jun 19 08:14:13 2024
*nat
:PREROUTING ACCEPT [0:0]
:INPUT ACCEPT [0:0]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [1:96]
-A POSTROUTING -s 10.66.27.0/24 -o enp2s0 -m comment --comment wireguard-nat-rule -j MASQUERADE
COMMIT
# Completed on Wed Jun 19 08:14:13 2024

```

**Reply by mwoolweaver** [reply]
it seems that iptables-persistent is not working properly and so the firewall rules are not staying when you reboot. i mentioned somethings above about possible solutions. 

https://github.com/pivpn/pivpn/discussions/1845#discussioncomment-9810895

**Comment by orazioedoardo** [comment]
> I had to change from "Allow only local requests" to "Respond Only On Interface enp2s0" because I was not able to browse the internet with the initial setting.

Did you do this AFTER installing PiVPN? If you did, you can’t use Pi-hole over VPN.

**Reply by joseb552** [reply]
Oh I might have, don't 100% remember the order of operations but it's likely I did it after installing PiVPN. 

Should I wipe and then install PiHole, make that change and THEN install PiVPN? Or is that setting incompatible with PiVPN? I'm honestly not even sure why I had to make that change to begin with. The PC is in the same environment as the Pi was and that never had any issues with the "Allow only local requests" setting.

**Reply by orazioedoardo** [reply]
PiVPN changes listening behavior to allow local requests so that DNS from VPN works. You don’t need to reinstall for this, just change to "Allow only local requests" in the web interface and apply.

**Reply by joseb552** [reply]
Ok I'll give that a shot. I initially changed it because when I switched the router to use the new system as DNS (no VPN), it wasn't processing any requests and the only way I could get it to process was by changing it.

**Reply by joseb552** [reply]
Yeah I'm getting the same issue from before. When I enable "Allow only local requests" and I change my iPhone's DNS entry to this PiHole, I get this error and it won't load anything on the phone / device. 

```
*** [ DIAGNOSING ]: Pi-hole diagnosis messages
   count   last timestamp       type                  message                                                       blob1                 blob2                 blob3                 blob4                 blob5               
   ------  -------------------  --------------------  ------------------------------------------------------------  --------------------  --------------------  --------------------  --------------------  --------------------
   1       2024-06-18 17:01:52  DNSMASQ_WARN          ignoring query from non-local network 192.168.1.157 (logged                                                                                                               
                                                      only once)             
```

**Reply by orazioedoardo** [reply]
That’s weird, try this: first change to “Listen only on interface enp2s0”, then create the file `/etc/dnsmasq.d/0N-listen-pivpn.conf` with content `interface=wg0`. **N** in the filename should be one greater or more than the filename in `/etc/dnsmasq.d` with the biggest number, so say the it’s `03-pihole-dhcp.conf`, then you would create `/etc/dnsmasq.d/04-listen-pivpn.conf`. Restart pi-hole.

**Reply by joseb552** [reply]
I changed the setting to 'listen only on interface enp2s0', then created the file as you said. 

```
odge@odge-pihole:/etc/dnsmasq.d$ ls
01-pihole.conf  02-pivpn.conf  06-rfc6761.conf  07-listen-pivpn.conf
odge@odge-pihole:/etc/dnsmasq.d$ more 07-listen-pivpn.conf
interface=wg0
```

Then I rebooted the system, when it was back up I ran pivpn -d and hit 'Y' to both issues it found (same as before). Then I switched my router to use the new system and tested on my phone (disconnected from WiFi, connected to VPN) and it seems to be working correctly now. I was able to visit some websites from the phone using VPN. 

I'll leave this running in this configuration while I head out to work for the day. I'll see if it holds after a reboot later.

**Reply by orazioedoardo** [reply]
Regarding the iptables issue, there’s my reply above asking for more commands to diagnose.

**Reply by joseb552** [reply]
Ah sorry, missed that previous request. Just updated with outputs.
